[Login] Add 2FA components to wp-login

Tug opened the pull request:
This PR is based on Luna's work https://github.com/Automattic/wp-calypso/pull/11367.
It adds 2FA components to our new `login`.

### Testing Instructions
- Apply patch D5442-code
- Checkout this branch and run app locally
- Go to http://calypso.localhost/login from a logged out session
- Try logging in with an account which does not have 2FA
- Logout and go back to the login page
- Try logging in with an account which has 2FA, if you have the app installed you should see the mobile push screen, otherwise you should see the 2FA code screen
- Change this line to simulate other types of 2FA https://github.com/Automattic/wp-calypso/pull/13235/files#diff-be252b9b0566b7767ed4492c5c873bb1R26

### Reviews
- [ ] Product
- [ ] Code

### Known issues
- [ ] When using Chrome password manager, username and password are emptied out when we click on "Log in" (may fix)
- [x] Visiting `/login` while logged in makes this screen fail (https://github.com/Automattic/wp-calypso/pull/13212 should fix that)
- [ ] Hyperlinks in the 2FA mobile push screen do not do anything, they would need to be connected to some UI state but I found this PR to be large enough as it is (wont fix now)
- [ ] None of the 2FA work currently, need follow up server patches (wont fix now)
- [ ] Back button is broken when seeing the 2FA components (we would expect to go back to login)
- [ ] Username and Password sometimes appears as get parameters in the url. I think this is due to `/login` being server-side rendered when we're already logged in, would need to investigate

a8ci18n commented on client/blocks/login/two-factor-authentication/verification-code-input.jsx:
```diff
@@ -0,0 +1,149 @@
+				<form onSubmit={ this.onCodeSubmit }>
+					<Card>
+						<p>
+							{ this.props.translate( 'Please enter the verification code generated' +
```
Hi! I've found a possible matching string that has already been translated **27** times:
`translate( 'Please enter the verification code generated by your Authenticator mobile application.' )` *ES Score: 8*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

a8ci18n commented on client/blocks/login/two-factor-authentication/waiting-notification-approval.jsx:
```diff
@@ -0,0 +1,48 @@
+							{ this.props.translate( 'Or, continue to your account using:' ) }
+						</p>
+						<p>
+							<a href="#">{ this.props.translate( 'A recovery code via text' ) }</a>
```
Hi! I've found a possible matching string that has already been translated **40** times:
`translate( 'Send recovery code via text' )` *ES Score: 12*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

a8ci18n commented on client/blocks/login/two-factor-authentication/waiting-notification-approval.jsx:
```diff
@@ -0,0 +1,48 @@
+				<form onSubmit={ this.onSubmitForm }>
+					<Card className="is-compact">
+						<p>
+							{ this.props.translate( 'We just sent a push notification to your WordPress mobile app. ' +
```
Hi! I've found a possible matching string that has already been translated **6** times:
`translate( 'We just sent a push notification to your <strong>WordPress mobile app</strong>. Swipe or tap to open and verify your login.' )` *ES Score: 7*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

a8ci18n commented on client/blocks/login/two-factor-authentication/verification-code-input.jsx:
```diff
@@ -0,0 +1,148 @@
+				<form onSubmit={ this.onCodeSubmit }>
+					<Card>
+						<p>
+							{ translate( 'Please enter the verification code generated' +
```
Hi! I've found a possible matching string that has already been translated **27** times:
`translate( 'Please enter the verification code generated by your Authenticator mobile application.' )` *ES Score: 8*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

a8ci18n commented on client/blocks/login/two-factor-authentication/waiting-notification-approval.jsx:
```diff
@@ -0,0 +1,50 @@
+							{ translate( 'Or, continue to your account using:' ) }
+						</p>
+						<p>
+							<a href="#">{ translate( 'A recovery code via text' ) }</a>
```
Hi! I've found a possible matching string that has already been translated **40** times:
`translate( 'Send recovery code via text' )` *ES Score: 12*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

a8ci18n commented on client/blocks/login/two-factor-authentication/waiting-notification-approval.jsx:
```diff
@@ -0,0 +1,50 @@
+				<form onSubmit={ this.onSubmitForm }>
+					<Card className="is-compact">
+						<p>
+							{ translate( 'We just sent a push notification to your WordPress mobile app. ' +
```
Hi! I've found a possible matching string that has already been translated **6** times:
`translate( 'We just sent a push notification to your <strong>WordPress mobile app</strong>. Swipe or tap to open and verify your login.' )` *ES Score: 7*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

Tug replied:
Sadly we can't use it due to the html tags in the string

Tug commented on client/blocks/login/index.jsx:
```diff
@@ -3,148 +3,56 @@
  */
```
Ignore the diff on this file, just read the code as a whole. Its previous content was basically moved to `client/blocks/login/login-form.jsx`

a8ci18n commented on client/blocks/login/two-factor-authentication/waiting-notification-approval.jsx:
```diff
@@ -0,0 +1,50 @@
+				<form onSubmit={ this.onSubmitForm }>
+					<Card className="two-factor-authentication__push-notification-screen is-compact">
+						<p>
+							{ translate( 'We just sent a push notification to your WordPress mobile app. ' +
```
Hi! I've found a possible matching string that has already been translated **6** times:
`translate( 'We just sent a push notification to your <strong>WordPress mobile app</strong>. Swipe or tap to open and verify your login.' )` *ES Score: 7*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

Tug commented:
Let's move forward on this. I'm going to split this PR into several smaller ones.
Here is the first one https://github.com/Automattic/wp-calypso/pull/13526

Tug commented:
Rebased after #13526 was merged.

I also removed a lot of state related functions (reducers, selectors, actions, tests) that were not essentials for this first batch. We can use the promise returned by `loginUserWithTwoFactorVerificationCode` instead which makes the code much simpler to reason with as we did in #13026.

We will need to refactor those components anyway because all the login screens are on 1 route currently and this can be a problem if user click on Back for instance.

Finally, the `/wp-login.php` ajax login for 2FA is still being developed so we will need to update the code when it's merged.

drewblaisdell commented on client/blocks/login/index.jsx:
```diff
@@ -18,6 +20,11 @@ class Login extends Component {
 
+	state = {
+		usernamePasswordValid: false,
+		rememberMe: false,
```
Pretty minor, but what do you think about making these property names a little clearer? Something like `hasSubmittedValidCredentials` and `isRememberMeChecked`?

drewblaisdell commented on client/blocks/login/two-factor-authentication/verification-code-input.jsx:
```diff
@@ -0,0 +1,119 @@
+		let errorText = this.state.error;
+		if ( isError ) {
+			errorText = translate( 'Invalid verification code' );
+		}
```
I don't see a reason to assign `errorText` to `this.state.error` at any point if we always use the same message. What do you think about using the following?

```js
const isError = !! this.state.error;
const errorText = isError && translate( 'Invalid verification code' );
```

drewblaisdell commented on client/blocks/login/two-factor-authentication/verification-code-input.jsx:
```diff
@@ -0,0 +1,119 @@
+
+	state = {
+		twoStepCode: '',
+		error: null
```
When is `error` ever set to a non-null value? We don't have a `catch` handler for the `loginUserWithTwoFactorVerificationCode` promise, so it does not seem to me that this ever happens.

stephanethomas replied:
Agree. we should remove that.

drewblaisdell commented on client/state/login/selectors.js:
```diff
@@ -0,0 +1,57 @@
+	}
+
+	return (
+		twoFactorAuth &&
```
There is no need to do a presence check for this value here because there is one on line 48.

drewblaisdell commented on client/blocks/login/two-factor-authentication/waiting-notification-approval.jsx:
```diff
@@ -0,0 +1,50 @@
+	}
+}
+
+export default connect( null, {} )( localize( WaitingTwoFactorNotificationApproval ) );
```
Do we need `connect` here?

drewblaisdell commented:
I added some commits to address my minor comments. Please remove them if you disagree or squash them into yours if you agree. :)

Also, I think we need to have some discussion about the API return format in D5238-code before we can really proceed with the review here, as changes to that patch have e.g. made it impossible to see the push notification screen.

Tug commented:
Updated so we use the last version of D5238 when testing.

stephanethomas commented on client/blocks/login/two-factor-authentication/verification-code-input.jsx:
```diff
@@ -0,0 +1,117 @@
+				[ event.target.name ]: event.target.checked
+			} );
+			return;
+		}
```
We don't need that block, there is no checkbox in the form.

stephanethomas commented on client/state/login/actions.js:
```diff
@@ -24,6 +24,16 @@ const loginErrorMessages = {
+		return errorKeys.map( errorKey => loginErrorMessages[ errorKey ] ).join( ' ' );
+	}
+
+	return get( error, 'response.body.data' ) || error.message;
```
I think you should be able to do this:

```javascript
return get( error, 'response.body.data', error.message );
```

stephanethomas commented on client/state/login/actions.js:
```diff
@@ -45,26 +55,49 @@ export const loginUser = ( usernameOrEmail, password ) => dispatch => {
+				two_step_nonce,
+				remember,
+				client_id: config( 'wpcom_signup_id' ),
+				client_secret: config( 'wpcom_signup_key' ),
```
I think some of those parameters are incorrect. The new login endpoint expects `user_id` and `remember_me`.

stephanethomas commented on client/state/login/selectors.js:
```diff
@@ -0,0 +1,56 @@
+ * @return {?Number}         The user ID.
+ */
+export const getTwoFactorAuthId = ( state ) => {
+	return get( state, [ 'login', 'twoFactorAuth', 'two_step_id' ], null );
```
Why did you prefer using an array for describing the path to this value instead of a string here and below? The following looks much shorter and readable to me:

```javascript
return get( state, 'login.twoFactorAuth.two_step_id', null );
```

jblz commented on client/state/login/test/reducer.js:
```diff
@@ -16,7 +16,8 @@ import {
 	requestError,
-	requestSuccess
+	requestSuccess,
+	twoFactorAuth
```
Trailing comma, please :)

jblz commented on client/state/login/test/selectors.js:
```diff
@@ -0,0 +1,105 @@
+				}
+			} );
+
+			expect( id ).to.eql( 123456 );
```
I'd use `.equal` instead of `.eql` here for a strict equality check.

jblz commented on client/state/login/test/selectors.js:
```diff
@@ -0,0 +1,105 @@
+				}
+			} );
+
+			expect( nonce ).to.eql( 'abcdef123456' );
```
I'd use `.equal` instead of `.eql` here for a strict equality check.

drewblaisdell commented on client/blocks/login/two-factor-authentication/verification-code-input.jsx:
```diff
@@ -0,0 +1,117 @@
+		const errorText = isError && translate( 'Invalid verification code' );
+
+		return (
+			<div>
```
We do, because there are siblings of `<form>`.

stephanethomas replied:
Ops, I missed that part in the diff.

drewblaisdell commented on client/state/login/actions.js:
```diff
@@ -45,26 +55,49 @@ export const loginUser = ( usernameOrEmail, password ) => dispatch => {
+ */
+export const loginUserWithTwoFactorVerificationCode = ( two_step_id, two_step_code, two_step_nonce, remember ) => {
+	return () => {
+		return request.post( config( 'login_url_xhr' ) )
```
This uses a separate endpoint. :)

drewblaisdell commented:
I updated this to address most of the notes above, as well as fix a couple bugs, use the new endpoint for the 2FA request, and use the new properties we decided on elsewhere.

Currently, this works until you are asked to submit a 2FA code: whether you are using the mobile app or SMS/Authenticator, you should see the correct component based on the response from the API. Neither submitting a code or continuing after approving the notification in the mobile app is implemented yet. I believe, given the size of this PR, that we should handle these separately.

Tug commented on client/blocks/login/index.jsx:
```diff
@@ -18,6 +20,11 @@ class Login extends Component {
+	state = {
+		hasSubmittedValidCredentials: false,
+		rememberMe: false,
+	};
```
See https://github.com/Automattic/wp-calypso/pull/13235#issuecomment-298366685
We might switch to a redux state in a later batch.

Tug commented:
LGTM!
I'll let an external reviewer (@stephanethomas?) give the final 👍

yurynix commented on client/state/login/actions.js:
```diff
@@ -45,26 +55,47 @@ export const loginUser = ( usernameOrEmail, password ) => dispatch => {
+			client_secret: config( 'wpcom_signup_key' ),
+		} )
+		.then( () => { /* TODO: Handle successful request */ } )
+		.catch( () => { /* TODO: Handle error */ } );
```
Minor we can fix later:
This swallows the error, so even on error it will go to success

yurynix commented:
It works for me as described 👍 , there are some more things to do but we can handle that later.
